pkg/dbal/DbalConsumer.php

```diff
@@ -66,41 +74,30 @@ public function getQueue(): Queue
         return $this->queue;
     }
 
-    public function receive(int $timeout = 0): ?Message
+    public function receiveNoWait(): ?Message
     {
-        $timeout /= 1000;
-        $startAt = microtime(true);
-
-        while (true) {
-            $message = $this->receiveMessage();
-
-            if ($message) {
-                return $message;
-            }
-
-            if ($timeout && (microtime(true) - $startAt) >= $timeout) {
-                return null;
-            }
+        $redeliveryDelay = $this->getRedeliveryDelay() / 1000; // milliseconds to seconds
 
-            usleep($this->pollingInterval * 1000);
+        $this->redeliverMessages();
 
-            if ($timeout && (microtime(true) - $startAt) >= $timeout) {
-                return null;
+        // get top message from the queue
+        if ($message = $this->fetchMessage([$this->queue->getQueueName()], $redeliveryDelay)) {
+            if ($message['redelivered'] || empty($message['time_to_live']) || $message['time_to_live'] > time()) {
```

Comment:
the expired message has to be deleted from message, I think that expiration logic could be moved to fetchMessage method

Reply:
I gave it another though, and I now I think we could cook a query that removes expired messages. just like redeliverMessage. the expiration in the code should still be preserved